### PR TITLE
redux migration

### PR DESCRIPTION
**Pull request by ostannya**

### CODE REVIEW

File: src/components/Home.js [hunk 1]
```diff
@@ -30,77 +51,60 @@ export class Home extends React.Component {
...
   }
 
-  handleCompare () {
-    const original = this.state.original
-    const changed = this.state.changed
-    const diffArray = Diff.diffChars(original, changed)
+  handleCompare (original, changed) {
     if (original === changed) {
-      this.setState({ identical: true, result: false })
+      store.dispatch(identical())
```
Comment (delfrrr) [inline src/components/Home.js:58]: 1. One event handler -> one action
2. No implicit dependencies between actions
3. If one action triggers another action, then thunk must be used

File: src/components/Home.js [hunk 2]
```diff
@@ -117,23 +91,23 @@ export class Home extends React.Component {
...
                     <Button className={styles.clearButton} onClick={this.handleClear}>Clear</Button>
                     <Tooltip title='Swap'>
-                      <SwapOutlined className={styles.swapButton} onClick={this.handleSwap} />
+                      <SwapOutlined className={styles.swapButton} onClick={() => this.handleSwap(original, changed)} />
                     </Tooltip>
                   </div>
                 </div>)
           }
           <Input
             onChange={this.handleChange}
```
Comment (delfrrr) [inline src/components/Home.js:107]: Does not do anything

File: src/components/Home.js [hunk 3]
```diff
@@ -6,101 +6,75 @@ import Toolbar from './Toolbar'
...
-    } else {
-      this.setState({ diffArray: diffArray, result: true, identical: false })
-    }
+    store.dispatch(compare(this.props.original, this.props.changed))
   }
 
-  handleChange (original, changed) {
-    this.setState({ original, changed })
+  handleChange () {
+    store.dispatch(valueChange(this.props.original, this.props.changed))
```
Comment (delfrrr) [inline src/components/Home.js:49]: does not do anything and would not work as uses previous values

File: src/components/Input.js [hunk 4]
```diff
@@ -15,13 +25,11 @@ export class Input extends React.Component {
   }
 
   handleChange () {
-    this.setState({ original: this.original.current.value, changed: this.changed.current.value })
-    this.props.onChange(this.original.current.value, this.changed.current.value)
+    store.dispatch(valueChange(this.original.current.value, this.changed.current.value))
```
Comment (delfrrr) [inline src/components/Input.js:28]: Does not need referrer in this component

**Review by delfrrr — CHANGES_REQUESTED** [review]

File: package.json [hunk 5]
```diff
@@ -6,6 +6,7 @@
   "dependencies": {
     "@ant-design/icons": "^4.5.0",
     "@jamesives/github-pages-deploy-action": "^4.1.1",
+    "@reduxjs/toolkit": "^1.6.0",
```
Comment (delfrrr) [inline package.json:9]: 🤔

File: src/components/Home.js [hunk 6]
```diff
@@ -117,23 +85,22 @@ export class Home extends React.Component {
               : (
                 <div>
                   <Toolbar
-                    diffArray={this.state.diffArray} onMenuClick={this.handleMenuClick} onLowercase={this.handleLowercase}
+                    diffArray={this.props.diffArray} onMenuClick={this.handleMenuClick} onLowercase={this.handleLowercase}
```
Comment (delfrrr) [inline src/components/Home.js:88]: * `Toolbar` should not depend on props which can be taken from `store`
* it makes sense to incapsulate event handlers into the `Toolbar`

Comment (delfrrr) [reply src/components/Home.js:88]: Same for elements below

File: src/components/Input.js [hunk 7]
```diff
@@ -1,27 +1,31 @@
...
 }
 
-export class Input extends React.Component {
+class Input extends React.Component {
   constructor (props) {
     super(props)
-    this.handleChange = this.handleChange.bind(this)
-    this.changed = React.createRef()
-    this.original = React.createRef()
+    this.handleOriganalChange = this.handleOriganalChange.bind(this)
```
Comment (delfrrr) [inline src/components/Input.js:15]: not nececerly

File: src/redux/actions.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    type: VALUE_CHANGE,
+    original,
+    changed
+  }
+}
+
+function diffCompare (original, changed) {
+  const diffArray = Diff.diffChars(original, changed)
+  return {
+    type: DIFF_ARRAY,
```
Comment (delfrrr) [inline src/redux/actions.js:44]: action type and action creator naming mismatch

File: src/redux/actions.js [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  return (dispatch) => {
+    if (original === changed) {
+      dispatch(identical())
+      dispatch(noResult())
+      dispatch({ type: COMPARE })
+    } else {
+      dispatch(diffCompare(original, changed))
+      dispatch(notIdentical())
+      dispatch(result())
+      dispatch({ type: COMPARE })
```
Comment (delfrrr) [inline src/redux/actions.js:59]: code duplication

File: src/redux/actions.js [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    dispatch(noResult())
+    dispatch({ type: CLEAR })
+  }
+}
+
+export function lowercase () {
+  return (dispatch, getState) => {
+    const { original, changed } = getState()
+    dispatch(valueChange(original.toLocaleLowerCase(), changed.toLocaleLowerCase()))
+    dispatch(notIdentical())
```
Comment (delfrrr) [inline src/redux/actions.js:86]: fasle logic

File: src/redux/reducers.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  IS_IDENTICAL,
+  IS_NOT_IDENTICAL,
+  RESULT,
+  NO_RESULT,
+  VALUE_CHANGE,
+  DIFF_ARRAY
+} from './constants.js'
+
+export function identical (state = false, action) {
+  switch (action.type) {
```
Comment (delfrrr) [inline src/redux/reducers.js:11]: identical logic should be here

File: src/components/Input.js (hunk 7 above)
Comment (ostannya) [reply src/components/Input.js:15]: + this.handleCompare in Home.js

File: src/redux/reducers.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    case VALUE_CHANGE:
+      return action.original === action.changed
+    case NOT_IDENTICAL:
+      return false
+    default:
+      return state
+  }
+}
+
+export function result (state = false, action) {
```
Comment (delfrrr) [inline src/redux/reducers.js:20]: unclear name; for example `showResult` is better

File: src/redux/reducers.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+  switch (action.type) {
+    case RESULT:
```
Comment (delfrrr) [inline src/redux/reducers.js:22]: `RESULT` and `NO_RESULT` does not represent any event or user intent

File: src/redux/actions.js [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    const { original, changed } = getState()
+    dispatch(valueChange(changed, original))
+    dispatch(compare(changed, original))
+    dispatch({ type: SWAP })
+  }
+}
+
+export function clear () {
+  return (dispatch) => {
+    dispatch(valueChange('', ''))
```
Comment (delfrrr) [inline src/redux/actions.js:66]: Logic should be in reducer

File: src/redux/actions.js [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+export function valueChange (original, changed) {
+  return {
+    type: VALUE_CHANGE,
+    original,
+    changed
+  }
+}
+
+function diffCompare (original, changed) {
```
Comment (delfrrr) [inline src/redux/actions.js:37]: what's different with `compare`?

File: src/components/Toolbar.js [hunk 16]
```diff
@@ -1,16 +1,42 @@
...
 export class Toolbar extends React.Component {
+  constructor (props) {
+    super(props)
+    this.handleLowercase = this.handleLowercase.bind(this)
+    this.handleBreaksToSpaces = this.handleBreaksToSpaces.bind(this)
+    this.handleRemoveWhiteSpaces = this.handleRemoveWhiteSpaces.bind(this)
+  }
+
+  handleLowercase () {
+    store.dispatch(lowercase(this.props.original, this.props.changed))
```
Comment (ostannya) [inline src/components/Toolbar.js:21]: does not need to pass variables which can be taken frm store

File: src/redux/actions.js [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+export function valueChange (original, changed) {
+  return {
+    type: VALUE_CHANGE,
+    original,
+    changed
+  }
+}
+
+export function compare (original, changed) {
```
Comment (ostannya) [inline src/redux/actions.js:20]: does not require input params which can be taken from store

**Review by delfrrr — APPROVED** [review]

**Pull request merged**